samples/net/mqtt/src/modules/transport/transport.c

```diff
@@ -0,0 +1,344 @@
+		.device_id.ptr = client_id,
+		.device_id.size = strlen(client_id),
+	};
+
+	err = mqtt_helper_connect(&conn_params);
+	if (err) {
+		LOG_ERR("Failed connecting to MQTT, error code: %d", err);
+	}
+
+	k_work_reschedule(&connect_work,
+			  K_SECONDS(CONFIG_MQTT_SAMPLE_TRANSPORT_RECONNECTION_TIMEOUT_SECONDS));
+}
+
+/* Zephyr State Machine framework handlers */
+static void disconnected_entry(void *o)
+{
+	struct s_object *od = o;
+
+	/* Reschedule a connection attempt if we are connected to network and we enter the
+	 * disconnected state.
+	 */
+	if (od->status == NETWORK_CONNECTED) {
+		k_work_reschedule_for_queue(&queue, &connect_work, K_NO_WAIT);
+	}
+}
+
+static void disconnected_run(void *o)
+{
+	struct s_object *od = o;
+
+	if ((od->status == NETWORK_DISCONNECTED) && (od->chan == &NETWORK_CHAN)) {
+		/* If NETWORK_DISCONNECTED is received after the MQTT connection is closed,
+		 * we cancel the connect work if it is onging.
+		 */
+		k_work_cancel_delayable(&connect_work);
+	}
+
+	if ((od->status == NETWORK_CONNECTED) && (od->chan == &NETWORK_CHAN)) {
+
+		/* Wait for 5 seconds to ensure that the network stack is ready before
+		 * attempting to connect to MQTT. This delay is only needed when building for
+		 * Wi-Fi.
+		 */
+		k_work_reschedule_for_queue(&queue, &connect_work, K_SECONDS(5));
+	}
+}
+
+static void connected_entry(void *o)
+{
+	LOG_INF("Connected to MQTT broker");
+	LOG_INF("Hostname: %s", CONFIG_MQTT_SAMPLE_TRANSPORT_BROKER_HOSTNAME);
+	LOG_INF("Client ID: %s", client_id);
+	LOG_INF("Port: %d", CONFIG_MQTT_HELPER_PORT);
+	LOG_INF("TLS: %s", IS_ENABLED(CONFIG_MQTT_LIB_TLS) ? "Yes" : "No");
+
+	ARG_UNUSED(o);
+
+	/* Cancel any ongoing connect work when we enter connected state */
+	k_work_cancel_delayable(&connect_work);
+
+	subscribe();
+}
+
+static void connected_run(void *o)
+{
+	struct s_object *od = o;
+
+	if ((od->status == NETWORK_DISCONNECTED) && (od->chan == &NETWORK_CHAN)) {
+		/* Explicitly disconnect the MQTT transport when losing network connectivity.
+		 * This is to cleanup any internal library state.
+		 * The call to this function will cause on_mqtt_disconnect() to be called.
+		 */
+		(void)mqtt_helper_disconnect();
+		return;
+	}
+
+	if (od->chan != &PAYLOAD_CHAN) {
+		return;
+	}
+
+	publish(&od->payload);
+}
+
+static void connected_exit(void *o)
+{
+	ARG_UNUSED(o);
+
+	LOG_INF("Disconnected from MQTT broker");
+}
+
+/* Construct state table */
+static const struct smf_state state[] = {
+	[MQTT_DISCONNECTED] = SMF_CREATE_STATE(disconnected_entry, disconnected_run, NULL),
+	[MQTT_CONNECTED] = SMF_CREATE_STATE(connected_entry, connected_run, connected_exit),
+};
+
+static void transport_task(void)
+{
+	const struct zbus_channel *chan;
+	enum network_status status;
+	struct payload payload;
+	int err;
+
+#if defined(CONFIG_MODEM_KEY_MGMT)
+	err = credentials_provision();
+	if (err) {
+		LOG_ERR("credentials_provision, error: %d", err);
+		return;
```

Comment:
This is irrecoverable, and I think that needs to be notified more clearly, and possiby lead to some action